src/main/java/io/reactivex/internal/operators/observable/ObservableDistinct.java

```diff
@@ -76,15 +51,33 @@ public boolean test(K t) {
             }
         };
         
-        return new NbpOperatorDistinct<T, K>(keySelector, p);
+        return new ObservableDistinct<T, K>(source, keySelector, p);
     }
     
-    @SuppressWarnings("unchecked")
-    public static <T> NbpOperatorDistinct<T, T> untilChanged() {
-        return (NbpOperatorDistinct<T, T>)UNTIL_CHANGED;
+    public static <T> ObservableDistinct<T, T> untilChanged(ObservableConsumable<T> source) {
+        Supplier<? extends Predicate<? super T>> p = new Supplier<Predicate<T>>() {
+            @Override
+            public Predicate<T> get() {
+                final Object[] last = { null };
```

Comment:
It wasn't an optimization target in this PR but I'll try to remember it when I add op-fusion (this could disappear completely).